optimum/intel/neural_compressor/trainer.py

```diff
@@ -553,17 +553,18 @@ def compute_loss(self, model, inputs, return_outputs=False):
             self.is_in_train
             and hasattr(self, "agent")
             and hasattr(self.agent, "criterion")
-            and hasattr(self.agent.criterion, "teacher_model_forward")
+            and self.agent.criterion is not None
         ):
             logits = self._get_logits(outputs)
-            teacher_outputs = inputs.pop("teacher_outputs", None)
-            # Compute teacher model outputs
-            if teacher_outputs is None:
-                teacher_outputs = self.agent.criterion.teacher_model_forward(
-                    inputs, teacher_model=self.agent.teacher_model._model
-                )
-
-            teacher_logits = self._get_logits(teacher_outputs)
+            teacher_logits = inputs.pop("teacher_logits", None)
+            if hasattr(self.agent, "on_after_compute_loss"):
+                if teacher_logits is None:
+                    teacher_outputs = self.agent.criterion.teacher_model_forward(inputs)
+                    teacher_logits = self._get_logits(teacher_outputs)
+            elif hasattr(self.agent, "on_post_forward"):
+                # Compute teacher model outputs
+                self.agent.on_post_forward(inputs, teacher_output=teacher_logits)
+                teacher_logits = self._get_logits(self.agent.criterion.teacher_outputs)

```

Comment:
```suggestion
            # Compute teacher model outputs
            if teacher_logits is None:
                if hasattr(self.agent, "on_after_compute_loss"):
                    teacher_outputs = self.agent.criterion.teacher_model_forward(inputs)
                    teacher_logits = self._get_logits(teacher_outputs)
                elif hasattr(self.agent, "on_post_forward"):
                    self.agent.on_post_forward(inputs, teacher_output=teacher_logits)
                    teacher_logits = self._get_logits(self.agent.criterion.teacher_outputs)
                else:
                    raise ValueError("Unable to compute the teacher outputs")
```

Reply:
The on_after_compute_loss API usage is as above, so here is only a transitional change. in the future, we only use total_loss == agent.on_after_compute_loss(None, student_outputs, student_loss, teacher_outputs) directly.

Reply:
If teacher_logits is None, agent.on_post_forward will compute the tearcher_outputs, but if teacher_logits is not None, agent.on_post_forward didn't recompute teacher_outputs, just only assign to self.agent.criterion.teacher_outputs

Reply:
Oh, yes, I got your point now, your suggestion is OK.